mobile/src/full/java/org/openhab/habdroid/core/GcmListenerService.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2010-2018, openHAB.org and others.
+ *
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.openhab.habdroid.core;
+
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.net.Uri;
+import android.os.Bundle;
+import android.preference.PreferenceManager;
+import android.support.v4.app.NotificationCompat;
+import android.support.v4.content.ContextCompat;
+
+import org.openhab.habdroid.R;
+import org.openhab.habdroid.ui.OpenHABMainActivity;
+import org.openhab.habdroid.util.Constants;
+
+public class GcmListenerService extends com.google.android.gms.gcm.GcmListenerService {
+    static final String EXTRA_NOTIFICATION_ID = "notificationId";
+
+    @Override
+    public void onMessageReceived(String from, Bundle data) {
+        String notificationIdString = data.getString(EXTRA_NOTIFICATION_ID);
+        int notificationId = notificationIdString != null
+                ? Integer.parseInt(notificationIdString) : 1;
+
+        if ("notification".equals(data.getString("type"))) {
+            //for now we use google.sent_time as a time reference for our notifications as the gcm
+            //message does not contain the actual event time from the openhab instance at the moment,
+            //in case gcm time is also missing, we fall back to the reception time which may be delayed
+            long timestamp = data.getLong("google.sent_time", System.currentTimeMillis());
+            sendNotification(data.getString("message"), timestamp, notificationId);
+            // If this is hideNotification, cancel existing notification with its id
```

Comment:
I think that's obvious from the code, I would remove this comment. Also, the else-if block is not needed, you can do that in a separate if.

You can probably also extract the body of the first if branch into it's own method with a better name (maybe the comment is than not needed anymore, or at least not 3 lines). The same for the second if body, which could make the code easier to understand :)

Reply:
The second branch already *is* extracted to a separate method. It's only the comment that makes it more than 2 lines.

Reply:
hmm, you're right. Ok, it's not ideal, but I think it's ok :)